### PR TITLE
Libtest tests fixed

### PR DESCRIPTION
**Pull request by alefedor**
Fixed a problem with not all tests being executed.
Improved tests.
Added ```AbstractLinCheckTest``` for testing all LinCheck strategies

### CODE REVIEW

File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/Reporter.kt [hunk 1]
```diff
@@ -130,4 +164,17 @@ fun StringBuilder.appendIncorrectResults(scenario: ExecutionScenario, results: E
...
         append(uniteActorsAndResults(scenario.postExecution, results.postResults))
     }
+}
+
+private fun StringBuilder.appendExceptionStackTrace(e: Throwable) {
+    val sw = StringWriter()
+    val pw = PrintWriter(sw)
+    val reducedStackTrace = mutableListOf<StackTraceElement>()
+    for (ste in e.stackTrace) {
+        if (ste.className.startsWith("org.jetbrains.kotlinx.lincheck.runner.")) break
```
Comment (ndkoval) [inline lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/Reporter.kt]: Let's print the whole stack trace in DEBUG mode

File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/runner/InvocationResult.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * The instance of this class represents a result of Runner.run invocation.
+ *
+ * [isError] is true if we didn't get correct results.
+ * [results] are the results that we got or IllegalStateException if absent.
+ */
+sealed class InvocationResult {
+    abstract val isError: Boolean
+    open val results: ExecutionResult get() = throw IllegalStateException("There are no results, but requested")
```
Comment (ndkoval) [inline lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/runner/InvocationResult.kt]: You need them only for `ValueInvocationResult`

File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/runner/InvocationResult.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+/**
+ * The invocation returned regular result.
+ */
+class ValueInvocationResult(override val results: ExecutionResult) : InvocationResult() {
```
Comment (ndkoval) [inline lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/runner/InvocationResult.kt]: ValueInvocationResult -> CompletedInvocationResult

File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/runner/InvocationResult.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    override val isError get() = false
+}
+
+/**
+ * The invocation ended in a deadlock.
+ */
+class DeadLockInvocationResult(val threadDump: Map<Thread, Array<StackTraceElement>>) : InvocationResult() {
```
Comment (ndkoval) [inline lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/runner/InvocationResult.kt]: DeadLock -> Deadlock

File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/runner/InvocationResult.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import org.jetbrains.kotlinx.lincheck.execution.ExecutionResult
+import java.lang.RuntimeException
+
+/**
+ * The instance of this class represents a result of Runner.run invocation.
+ *
+ * [isError] is true if we didn't get correct results.
+ * [results] are the results that we got or IllegalStateException if absent.
+ */
+sealed class InvocationResult {
```
Comment (ndkoval) [inline lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/runner/InvocationResult.kt]: internal

File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt [hunk 6]
```diff
@@ -153,41 +153,34 @@ open class ParallelThreadsRunner(
...
-                    t as TestThread
-                    msgBuilder.appendln("Thread-${t.iThread}:")
-                    for (ste in stackTrace) {
-                        if (ste.className.startsWith("org.jetbrains.kotlinx.lincheck.runner.")) break
-                        msgBuilder.appendln("\t$ste")
-                    }
-                    msgBuilder.appendln()
+    override fun run(): InvocationResult {
+        try {
+            reset()
```
Comment (ndkoval) [inline lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt]: Should be before `try`

File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/strategy/IterationResult.kt [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+
+/**
+ * The instance of this class represents a result of Strategy.run usedInvocations.
+ *
+ * [scenario] is an ExecutionScenario on which the Strategy was runned.
+ * [usedInvocations] is a number of invocations used before we found a error or run of invocations.
+ * [isError] denotes whether a error occured during the iteration.
+ */
+sealed class IterationResult(val scenario: ExecutionScenario, val usedInvocations: Int) {
```
Comment (ndkoval) [inline lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/strategy/IterationResult.kt]: usedInvocations -> executedInvocations or just invocations

File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/strategy/IterationResult.kt [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ * The iteration got a deadlock.
+ */
+class DeadLockIterationResult(scenario: ExecutionScenario, usedInvocations: Int, val threadDump: Map<Thread, Array<StackTraceElement>>) : IterationResult(scenario, usedInvocations) {
+    override val isError get() = true
+}
+
+/**
+ * Converts InvocationResult with a error to an IterationResult with the corresponding error
+ */
+fun InvocationResult.toIterationResult(scenario: ExecutionScenario, usedInvocations: Int): IterationResult = when (this) {
```
Comment (ndkoval) [inline lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/strategy/IterationResult.kt]: asIterationResult

File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/strategy/Strategy.java [hunk 9]
```diff
@@ -46,19 +45,29 @@
...
         this.reporter = reporter;
     }
 
-    protected void verifyResults(ExecutionResult results) {
+    /**
+     * Check whether results are correct. Sets iterationResult if there was a error.
+     */
+    protected boolean verifyResults(InvocationResult result, int usedInvocations) {
+        if (result.isError()) {
+            iterationResult = IterationResultKt.toIterationResult(result, scenario, usedInvocations);
```
Comment (ndkoval) [inline lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/strategy/Strategy.java]: static import

File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/LinChecker.java [hunk 10]
```diff
@@ -74,38 +76,69 @@ public static void check(Class<?> testClass, Options options) {
         new LinChecker(testClass, options).check();
     }
 
+    /**
+     * Runs concurrent test on specified class with the specified by options environment.
+     *
+     * @return TestReport with information about concurrent test run.
+     */
+    static Map<CTestConfiguration, List<IterationResult>> analyze(Class<?> testClass, Options options) {
```
Comment (ndkoval) [inline lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/LinChecker.java]: Probably, `checkDetailed` or similar function name is better

File: libtest/src/test/java/org/jetbrains/kotlinx/lincheck/tests/AbstractLinCheckTest.kt [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * An abstraction for testing all lincheck strategies
+ */
+abstract class AbstractLinCheckTest(vararg expectedErrors: KClass<out IterationResult>) : VerifierState() {
+    private val expectedErrors = if (expectedErrors.isEmpty()) listOf(SuccessIterationResult::class) else expectedErrors.toList()
+    private val reporter = Reporter(LoggingLevel.INFO)
+
+    @Test
+    fun testStressStrategy() {
```
Comment (ndkoval) [inline libtest/src/test/java/org/jetbrains/kotlinx/lincheck/tests/AbstractLinCheckTest.kt]: testWithStressStrategy

File: libtest/src/test/java/org/jetbrains/kotlinx/lincheck/AnalyzeAccessor.kt [hunk 12]
```diff
@@ -0,0 +1,5 @@
+package org.jetbrains.kotlinx.lincheck
+
+import org.jetbrains.kotlinx.lincheck.strategy.IterationResult
+
+internal fun linCheckAnalyze(clazz: Class<*>, options: Options<*, *>?): Map<CTestConfiguration, List<IterationResult>> = LinChecker.analyze(clazz, options)
```
Comment (ndkoval) [inline libtest/src/test/java/org/jetbrains/kotlinx/lincheck/AnalyzeAccessor.kt]: It is better to make the method public and add a comment that this is an internal API. Don't want to create a new file for solving this issue, it makes the code more complicated

Comment (alefedor) [reply libtest/src/test/java/org/jetbrains/kotlinx/lincheck/AnalyzeAccessor.kt]: It seems to me that the fact that autocompletion suggests `checkDetailed` when typing `check` will attract unwanted attention

**Comment by ndkoval** [comment]
The changes were moved to #35, thanks for the cleaning up the code!

**Pull request closed without merging**